src/crimson/osd/osd_operation.h

```diff
@@ -379,12 +381,75 @@ class OperationThrottler : public Blocker,
   void release_throttle();
 };
 
+// the main template. by default an operation has no extenral
+// event handler (the empty tuple). specializing the template
+// allows to define backends on per-operation-type manner.
+// NOTE: basically this could be a function but C++ disallows
+// differentiating return type among specializations.
+template <class T>
+struct EventBackendRegistry {
+  template <typename...> static constexpr bool always_false = false;
+
+  static std::tuple<> get_backends() {
+    static_assert(always_false<T>, "Registry specializarion not found");
+    return {};
+  }
+};
+
+template <class T>
+constexpr static auto EVENT_NAMES = "UnknownEvent";
+
+void dump_event_default(const char name[],
+                        const utime_t& timestamp,
+                        ceph::Formatter* f);
+
+template <class T>
+struct Event {
+  // IDEA 0: should this be ExternalBackend? Or ExtraBackend?
+  struct Backend {
+    virtual void handle(T&, const Operation&) = 0;
+  };
+
+  struct InternalBackend final : Backend {
+    utime_t timestamp;
+    void handle(T&, const Operation&) override {
+      timestamp = ceph_clock_now();
+    }
+  } internal_backend;
+
+  // QUESTION: how to convey get_registered_backends() and glue it
+  // with particular Operation? The `events` are supposed to vary.
+  // QUESTION: do we need to handle any other Operation than Client
+  // Request when it comes to Historic?
```

Comment:
i think we also need to take `RepRequest` and `RecoverySubRequest` into account.